Applicative type class

felixwiemuth opened the pull request:
This adds an Applicative type class, including laws for Functor and Applicative as well as instances for `Option` and `List`.

Resolves #1649.

felixwiemuth commented:
For the status of instance implementations see #2186.

magnus-madsen commented on main/src/library/Applicative.flix:
```diff
@@ -0,0 +1,83 @@
+/// in the second applicative, where the possible quantity of functions/arguments depends on the type `m`.
+/// For example, an `Option[a -> b]` can be `None`, or contain a function of type `a -> b`, and only in the
+/// latter case a function is applied. A `List[a -> b]` contains a list of functions, which are all to be
+/// applied to all arguments in the arguments list.
```
I think one or two lines breaks in the above would be useful.

felixwiemuth replied:
Do you mean more paragraphs or shorter lines in general?

magnus-madsen replied:
Text is fine, I mean some literal line breaks, \n

felixwiemuth replied:
Yes, but to make more paragraphs out of the one paragraph or to have shorter lines?

magnus-madsen replied:
Breathing space:

```
/// A type class for functors that support application, i.e. allow to:
///
/// - Make an applicative value out of a normal value (embed it into a default context), e.g. embed `5` into `Some(5)`.
/// - Apply a function-type applicative to a matching argument-type applicative, resulting in an applicative of
///
/// the function's result type. The meaning of this application is defined by the respective instance.
```

magnus-madsen commented on main/src/library/Applicative.flix:
```diff
@@ -0,0 +1,83 @@
+/// they must be equivalent to their default definitions:
+///     `ap(f: m[a -> b & e], x: m[a]) : m[b] & e = liftA2(identity, f, x)`
+///     `liftA2(f: a -> b -> c & e, x: m[a], y: m[b]): m[c] & e = ap(Functor.map(f, x), y)`
+///
```
This of course means that you can forget to define either. What do you think about that?

felixwiemuth replied:
Yes that is of course a problem: if none is implemented, calling either will lead to non-termination.

For now I think we cannot do much more than document it if we want the flexibility of allowing to implement either one of the functions.

In the future we could introduce an advanced mechanism for type classes allowing to specify which functions have to be implemented. This could be a boolean expression composed of variables corresponding to the to-be-implemented functions. In this case the expression would be `ap v liftA2`.

magnus-madsen replied:
> In the future we could introduce an advanced mechanism for type classes allowing to specify which functions have to be implemented. This could be a boolean expression composed of variables corresponding to the to-be-implemented functions. In this case the expression would be `ap v liftA2`.

I like this idea.


But for now, how about we leave `ap` unimplemented? 

You can still write an instance that defines `ap` in terms of `lift` and overrides lift.

felixwiemuth replied:
> But for now, how about we leave ap unimplemented?
> You can still write an instance that defines ap in terms of lift and overrides lift.

Right, that sounds like the best option for now. Usually it is `ap` one wants to implement, and if one wants to implement `liftA2` instead, one can implement `ap` based on it like the default implementation does now. I could add that to the documentation.

magnus-madsen replied:
Agree. We do the same for e.g. `Order` where you must implement `compare`.

magnus-madsen commented on main/src/library/Applicative.flix:
```diff
@@ -0,0 +1,83 @@
+    /// Instances can define more efficient implementations than the default implementation
+    /// (which is `Applicative.ap(Functor.map(f, x), y)`).
+    ///
+    pub def liftA2(f: a -> b -> c & e, x: m[a], y: m[b]): m[c] & e = Applicative.ap(Functor.map(f, x), y)
```
What about liftAX? X> 2

felixwiemuth replied:
> Whats the plan for liftX?

Yes I guess it could be useful to have `liftAX` for X > 2, especially as this allows for more efficient implementations than the default implementation, which would be e.g. `liftA3(f, x1, x2, x3) = ap(liftA2(f, x1, x2), x3)` (or without indirections `ap(ap(map(f, x1), x2), x3)`).

For example, for the list instance, an implementation of `liftAX` which directly applies the function `f` to all combinations of the X arguments would be more efficient than creating a list of partially applied functions for each nested `ap` call (until reaching `liftA2` which might have the more efficient implementation).

For the `Option` instance, however, I guess the default implementation is the best one can do (except maybe for saving a few function calls).

One could also define lift for X=1, i.e. `liftA1(f: a -> b & e, x : m[a]) : m[b] = ap(point(f), x)`. But just as we have a law for `liftA2` requiring that `liftA2(f, x, y) == ap(map(f, x), y)`, we would require `liftA1(f,x) == map(f,x)`, so there is maybe no point in adding `liftA1`.

> Also the difference in naming between lift and liftA was not entirely clear to me.

I would say to distinguish applicative lift (`liftA`) from the corresponding monadic lift (`liftM`). Probably one eventually wants both in Prelude without namespacing prefix, and in the type classes where the instances are defined they of course also need to be distinguished.

It might still make sense to define `liftX` for e.g. `Option`, representing lifting of a function to work on `Option`s - here `liftX` would serve as implementation for both `liftAX` and `liftMX`.

magnus-madsen replied:
OK, so the actionable items are:

- [ ] Define liftXA in the TC. 
- [ ] Rename `Option.liftX` to `Option.liftXA` 

Does you agree? Do you want to do it in a separate PR?

felixwiemuth replied:
I would complete all the outstanding small changes in this PR.

I'm not sure about renaming e.g. `Option.liftX` to `Option.liftAX`. The argument was that this "lift" is also somewhat independent of Applicative or Monad. So I actually like that it is neutrally called `liftX`, while `liftAX` and `liftMX` will be defined based on that.

magnus-madsen replied:
OK and OK

magnus-madsen commented on main/src/library/Option.flix:
```diff
@@ -83,6 +83,13 @@ instance Functor[Option] {
+    override pub def ap(f: Option[a -> b & e], x: Option[a]) : Option[b] & e = Option.ap(f, x)
+    // Same as the default implementation using `map` and `ap` but with less indirection.
+    override pub def liftA2(f: a -> b -> c & e, x: Option[a], y: Option[b]): Option[c] & e = Option.lift2(f, x, y)
+}
```
More new lines in the  above.

magnus-madsen reviewed (commented):
This looks ready to me. I only have light stylistic issues.

Whats the plan for liftX?

Also the difference in naming between lift and liftA was not entirely clear to me. 

Did you see that in `Option.flix` there is a long list of lift operations?

felixwiemuth commented:
I should now have addressed all comments and have implemented `liftA3` through `liftA4` and added the corresponding implementation to `Option` and `List`.
